### PR TITLE
Feature: Add exists operator

### PR DESCRIPTION
**Pull request by mringler**
Did you realize, there is no possibility to do EXISTS in Propel, except through `->where('EXISTS (...))`, which has its own caveats (buggy, no query logic). So I had a look at how this probably should work, and it actually went ahead pretty quickly.
These changes allow to create exist queries in two ways:

First one is through regular useQuery logic, same as with the other high-level Propel statements. The relation condition is added to the inner query (in the example author.Id = book.author_id):  
```php
BookQuery::create()
->useAuthorExistsQuery()    // or ->useExistsQuery('Author')
  ->filterBy...
->endUse()
```

Second one is more basic, where a nested query is set as exists query. Here, any table can be used, but the filter in the exists-query must be added manually:
```php
$existsQuery = BookQuery::create()
->filterBy...
->where('Book.AuthorId = Author.Id'); // own join condition in where

AuthorQuery::create()
->whereExists($existsQuery)
```
As exists is such a basic part of SQL, I feel like there are lots of ways to use it which will not work with this implementation (yet). So it might need more refinement over time. Tests run through fine, but if you find anything, please let me know. Same goes for any feedback, of course.

### CODE REVIEW

**Comment by mringler** [comment]
Hmm, test  don't run through because import of BookQuery and Author query fails. Womp womp. Any suggestions?

File: tests/Propel/Tests/Runtime/ActiveQuery/ExistsTest.php
Comment (nederdirk) [inline tests/Propel/Tests/Runtime/ActiveQuery/ExistsTest.php:269]: A very minor remark: the test would tell its 'story' better, when the `$isGood` boolean parameter would be replaced with a `string $title`

**Review by nederdirk — COMMENTED** [review]
Hey @mringler, I wish this feature was available a few weeks ago (when I needed it in a project)! I have a few remarks, mostly regarding code clarity and readability. I'll leave it up to the maintainers to choose whether to require changes.

Thanks again!

**Review by mringler — COMMENTED** [review]
Hey @nederdirk, thank you for your review and your kind words! I agree with all your points, and make adjustments accordingly. Please let me know if there are other issues.

File: src/Propel/Runtime/ActiveQuery/Criterion/ExistsCriterion.php
Comment (mringler) [reply src/Propel/Runtime/ActiveQuery/Criterion/ExistsCriterion.php:31]: Yes, I agree. I thought about it, and then was admittedly too lazy to do it. Users shouldn't have to fumble with that anyway, but still. Interesting tip about the `@phpstan-var` parameter. Didn't know about that. Hope I am doing it right

File: src/Propel/Runtime/ActiveQuery/ModelCriteria.php
Comment (mringler) [reply src/Propel/Runtime/ActiveQuery/ModelCriteria.php:2104]: Yes, this is part of a necessary bugfix in another branch (https://github.com/propelorm/Propel2/pull/1704), and actually wasn't even part of the fix, but just cleanup on the way. I still think it was good to clean it up, but the way it ripples through to here is annoying, I agree.

File: src/Propel/Runtime/ActiveQuery/ModelCriteria.php
Comment (mringler) [reply src/Propel/Runtime/ActiveQuery/ModelCriteria.php:2179]: I agree once more. This is again part of the other bugfix, and was fixed there. I merged it in, hope this doesn't complicate further merges. Again, let me say just for relieve: it sucks to depend on an unmerged branch.

File: tests/Propel/Tests/Runtime/ActiveQuery/Criterion/ExistsCriterionTest.php
Comment (mringler) [reply tests/Propel/Tests/Runtime/ActiveQuery/Criterion/ExistsCriterionTest.php:74]: Yes, it comes from accessing the inner statement directly, and not through the ExistsFilter. When the ExistsFilter generates SQL, it sets a `SELECT 1 AS existsFlag`, but I wanted to test more isolated.
I would argue that the Criteria object should set the table columns by default, and never return incorrect SQL, but that is a different problem.

File: tests/Propel/Tests/Runtime/ActiveQuery/ExistsTest.php
Comment (mringler) [reply tests/Propel/Tests/Runtime/ActiveQuery/ExistsTest.php:269]: That would be more readable by itself, yes. In this particular case, I honestly like it better this way, as it makes the `createTestData()` method less cluttered, and allows you to see the whole data setup in just a glance, which was helpful when writing the tests. When you know what to look for, you can identify the number of "good" books in about a second.

Now that I have written the tests and don't rely on the short glace reminder anymore, I don't mind changing it to a more verbose signature, if you insist

File: src/Propel/Runtime/ActiveQuery/Criterion/ExistsCriterion.php
Comment (nederdirk) [inline src/Propel/Runtime/ActiveQuery/Criterion/ExistsCriterion.php:47]: Am I right that the `$type` parameter could be improved by having the default value be `self::TYPE_EXISTS` ? (looking at line 51 below)

File: src/Propel/Runtime/ActiveQuery/Criterion/ExistsCriterion.php
Comment (nederdirk) [inline src/Propel/Runtime/ActiveQuery/Criterion/ExistsCriterion.php:76]: This string-building operation can probably be simplified as such: `$sb .= $this->type . ' (' . $existsQuery . ')';`

File: tests/Propel/Tests/Runtime/ActiveQuery/ExistsTest.php
Comment (nederdirk) [reply tests/Propel/Tests/Runtime/ActiveQuery/ExistsTest.php:269]: The PHP interpreter doesn't mind either way, but it will save an interested developer/maintainer in a few months some clicks and thinking when scanning `createTestData`. In general I would say it's worth the effort to strive for maximum clarity

**Review by nederdirk — COMMENTED** [review]
Hey @mringler, I'm sorry for letting this wait so long. With your latest changes this PR is getting even better. With a few more things to think about or change we can probably land this plane!

File: src/Propel/Runtime/ActiveQuery/Criterion/ExistsCriterion.php
Comment (mringler) [reply src/Propel/Runtime/ActiveQuery/Criterion/ExistsCriterion.php:47]: Since the input parameter can be any string, I have to check if it is one of the expected ones anyway. Think of `new ExistsCriterion($outer, $inner, 'EXERSTST')`. So line 51 has to stay there anyway, and this comes down to if you think null is a better or worse default value. I think better, as it has less overhead.

File: src/Propel/Runtime/ActiveQuery/Criterion/ExistsCriterion.php
Comment (mringler) [reply src/Propel/Runtime/ActiveQuery/Criterion/ExistsCriterion.php:76]: Right, I'll change the TYPE_NOT_EXISTS so that this works

File: src/Propel/Generator/Builder/Om/QueryBuilder.php
Comment (mringler) [reply src/Propel/Generator/Builder/Om/QueryBuilder.php:1638]: Actually, it doesn't even work with a bool anymore, so I have to change it to the string input. Same for the generated useNotExists.

File: src/Propel/Runtime/ActiveQuery/Criterion/ExistsCriterion.php
Comment (mringler) [reply src/Propel/Runtime/ActiveQuery/Criterion/ExistsCriterion.php]: Thinking about it, having this depend on the type token being the correct SQL operator is not optimal, but now I have changed it already and it probably won't ever matter

File: tests/Propel/Tests/Runtime/ActiveQuery/Criterion/ExistsCriterionTest.php
Comment (mringler) [reply tests/Propel/Tests/Runtime/ActiveQuery/Criterion/ExistsCriterionTest.php:74]: Well, if said poor developer doesn't know how unfinished propel statements look like, this is probably as good a place as any to get familiar with it. I am more worried about the poor developer who repairs the ill behavior with the empty select columns, and has to update all the tests (try`grep -Hrn 'SELECT  FROM' tests/ | wc -l`). Though this probably won't happen.
On that note, I am curious, can you tell me how much Propel is still used? Looking at the state of the code, the state of the test cases, and how it seems to be stuck in 2.0@dev indefinitely, it seems like the project is on end-of-life support. Is that accurate? Would be a shame, I like a lot of it.

Comment (mringler) [reply tests/Propel/Tests/Runtime/ActiveQuery/Criterion/ExistsCriterionTest.php:74]: Interesting, thank you

**Review by nederdirk — APPROVED** [review]

File: src/Propel/Runtime/ActiveQuery/BaseModelCriteria.php
Comment (dereuromark) [inline src/Propel/Runtime/ActiveQuery/BaseModelCriteria.php:34]: We try to use same style for acronyms as for the rest of naming: `...InSql`
Is that possible here too?

File: src/Propel/Runtime/ActiveQuery/BaseModelCriteria.php
Comment (dereuromark) [inline src/Propel/Runtime/ActiveQuery/BaseModelCriteria.php:278]: Can we use non cloaking

    && $this->modelAlias

instead?

File: src/Propel/Runtime/ActiveQuery/Criterion/ExistsCriterion.php
Comment (dereuromark) [inline src/Propel/Runtime/ActiveQuery/Criterion/ExistsCriterion.php:82]: Is there a reason not to have this protected?

File: src/Propel/Runtime/ActiveQuery/BaseModelCriteria.php
Comment (dereuromark) [reply src/Propel/Runtime/ActiveQuery/BaseModelCriteria.php:34]: Oh, I see you moved this from somewhere else. So we better keep the same style then for now.

Comment (mringler) [reply src/Propel/Runtime/ActiveQuery/BaseModelCriteria.php:34]: Yes, I remember, that was a property defined in child class BaseModelCriteria, but then accessed in the parent class through `static`, and then stan complained about it (rightly so). And even worse, my editor couldn't find it, so I cleaned it up

File: src/Propel/Runtime/ActiveQuery/BaseModelCriteria.php
Comment (mringler) [reply src/Propel/Runtime/ActiveQuery/BaseModelCriteria.php]: Yes, good point. Changed it

File: src/Propel/Runtime/ActiveQuery/Criterion/ExistsCriterion.php
Comment (mringler) [reply src/Propel/Runtime/ActiveQuery/Criterion/ExistsCriterion.php]: Agreed, protected is better

**Pull request merged**
